Initial Petri nets version

adamburkegh opened the pull request:
Includes
* Petri nets implementation with label and weight support
* Petri net fragment parser implementation for Python
  * See the [Java implementation](https://github.com/adamburkegh/prom-helpers) and this [blog post](https://adamburkeware.net/2021/05/20/petri-net-fragments.html) for more context
* Export to DOT format and basic utility functions for invoking dot executable
* Export to PNML
* Unit tests

Does not include: 
* Exporter to pm4py objects or direct PNML visualisation support
  * I have functions for these, but did not want to include pm4py in the koalas dependencies
  * Should we put them in a `pmkoalas-pm4py` bridge project?

This code was originally implemented for [State Snapshot Miner](https://github.com/adamburkegh/statesnap-miner). I have changed the module names and removed a few functions that would bring in extra dependencies.

AdamBanham commented:
Hi Adam,

Could you provide some code snippets of how one might create a "Petri net"?

For example could you finish the following:
```python 
from pmkoalas.dtlog import convert
from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
from pmkoalas.models.petrinet import LabelledPetriNet

# create a log
log = convert( 
    "a b e d z",
    "a c e f z",
    "a y e g z"
)

# run alpha miner 
miner = AlphaMinerInstance()
places, transitions, flows = miner.mine_model(log)

# use given places, transitions and flows to create an 
# instance of a petri net 
net = LabelledPetriNet(places, transitions, flows)
# display net in some potential human readable manner
print(net)
```
However, running the above on the fork produces the following:
```
Traceback (most recent call last):
  File "F:\Projects\Koalas\working\net_test.py", line 20, in <module>
    print(net)
  File "F:\Projects\Koalas\pmkoalas\models\petrinet.py", line 59, in __str__
    return "LabelledPetriNet(" + self._label + ")\n" \
TypeError: can only concatenate str (not "NoneType") to str
```

AdamBanham commented:
Also I checked out the code, it seems mostly uncommented. We should at least strive to have the key entry points with a clear indication of how to use the function call.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+# during testing.
+#
+
+import operator
```
Why is this library imported? I cannot find where it has been used

adamburkegh replied:
Looks like it is unused.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+            +  " Arcs: " + str(self.arcs)
+    
+
+class MutableLabelledPetriNet(LabelledPetriNet):
```
This might seem a bit odd, but there is nothing private in python. Do we really need this class?

AdamBanham replied:
The functional extras that this class provides already exists, I believe.

AdamBanham replied:
```python
from pmkoalas.dtlog import convert
from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
from pmkoalas.models.petrinet import LabelledPetriNet

# create a log
log = convert( 
    "a b e d z",
    "a c e f z",
    "a y e g z"
)

# run alpha miner 
miner = AlphaMinerInstance()
places, transitions, flows = miner.mine_model(log)

# use given places, transitions and flows to create an 
# instance of a petri net 
net = LabelledPetriNet(places, transitions, flows, "hello")
# display net in some potential human readable manner
net._places.add("p3")
print(net)
```

AdamBanham replied:
In fact, I can do this in many ways, see below (very interesting) : 
```python

from pmkoalas.dtlog import convert
from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
from pmkoalas.models.petrinet import LabelledPetriNet

# create a log
log = convert( 
    "a b e d z",
    "a c e f z",
    "a y e g z"
)

# run alpha miner 
miner = AlphaMinerInstance()
places, transitions, flows = miner.mine_model(log)

# use given places, transitions and flows to create an 
# instance of a petri net 
net = LabelledPetriNet(places, transitions, flows, "hello")
# display net in some potential human readable manner
net._places.add("p3")
print(net)
print()

# or i could do this
places.add("p4")
print("printing initial set")
print(places)
print()
print("printing changed net")
print(net)
print()

# or this 
net.places.add("p5")
print(net)
```

adamburkegh replied:
Immutability (even if enforced by convention here) is useful in the implementation of the discovery algo. I didn't include it as I don't think obscure algos should be in the package. Specifically it uses frozensets at some points. I don't think mutability is a desirable property in a default PN impl.

AdamBanham replied:
You are reusing the references, adam; you have not enforced immutability by convention. Any collection in python is passed by reference, including sets. If you wish to enforce immutability then you must create a new collection in __init__ and all properties must return copies of the collections.

AdamBanham replied:
However, your comment below seems to indicate that the structure is pointless, other than to be a wrapper container. So why bother with any of these considerations? Just make it a pointless wrapper.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+    def __eq__(self,other):
+        if isinstance(other,LabelledPetriNet):
+            return self._label  == other._label and \
+                   self._places == other._places and \
```
Why do you need to reimplement this? The base class has this already

adamburkegh replied:
They differ in one line: `if isinstance(other,LabelledPetriNet):`, which allows MPNs to be compared with PNs

AdamBanham replied:
You can remove this function by checking that the other is a subclass of LabelledPetriNet in the top structure.
i.e. https://docs.python.org/3/library/functions.html#issubclass.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+                   self._arcs   == self._arcs
+        return False
+
+def verbosecmp(obj1,obj2):
```
I can see that you have used this function a lot in your tests but should this be in the library or in the testing module? Not really sure what this does

adamburkegh replied:
It gives a verbose description of differences between two of these objects. It is useful in a functional testing context as well as unit tests.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+
+
+
+class Place:
```
Why are these templates not used in typing for the structure? I can see why you talked about have a more modular structure for the net and its components. You have already added a lot of "extras" for the simple components of a petri net. I do agree with the need for a unique identifier, but perhaps we should handle that first and leave the option to set it off by default.

adamburkegh replied:
By template do you mean class? Do you mean in function signatures? It's mainly because I have trouble taking non-enforced typing seriously.

AdamBanham replied:
These classes do not have any functional behaviours other than to be holding data, I called these templates. Technically these can be dramatically reduced using dataclasses which will do all the work for you.

Type hinting allows for smoother construction, many will read at least the source for your function call. By having the hints there, a user can quicky find the appropriate class to use. The whole module already uses this pattern.

see https://docs.python.org/3/library/dataclasses.html

AdamBanham replied:
I can make you a decorator to enforce hints if you want, its not hard.

adamburkegh replied:
Templates is usually associated with generics because that is the jargon in C++.

These classes are likely to grow extra behaviour. I have used dataclasses in other places, but here they would indicate a design intent to keep these as simply records.

Runtime decorators to enforce types aren't high value.

I'm not against type hinting, it's fine, I will add it. But it is a bit silly without build-time tooling.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+
+
+SILENT_TRANSITION_DEFAULT_NAME='tau'
+def silentTransition(name=None,tid=None,weight=None):
```
I don't see how this would be helpful for a data structure. Can you tell me some more about this function?

adamburkegh replied:
You create a silent transition by calling this function, and you don't need to name it if you don't want to.

adamburkegh replied:
Eg, from state snapshot:
```
        tranId += 1
        tran = silentTransition(tid = tranId)
        atot[(fromPlaces,toPlaces)] = tran
        tweights[tran] = 1
```

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+        return dotstr
+
+
+def exportToDOT(net:LabelledPetriNet) -> str:
```
should be a function of the structure.

adamburkegh replied:
Hard disagree. Particular export formats need to be decoupled from the data structure. PNML is a standard for Petri nets which is the only reason it snuck under the wire. But the PNML functions could be moved out.

AdamBanham replied:
This only introduces pointless import statements and requires the user to know more pointless information about the hard structure of the module. This is very counter intuitive to me.

adamburkegh replied:
One thing I was considering was exporting some common functions across the packages in `__init__.py`. What are your thoughts on that idiom?

DOT comes with a load of external dependencies that really should stay out of the main data structure module. Same with pm4py, which I took out of the entire patch. Would like your thoughts about how to incorporate that, preferably without including pm4py as a dependency for koalas.

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+
+PNML_URL='http://www.pnml.org/version-2009/grammar/pnmlcoremodel',
+
+def exportToPNMLObj(net:LabelledPetriNet) -> ET.Element: 
```
Should be a function of the structure.

adamburkegh replied:
Big Ball of Mud waiting to happen

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+    return root
+
+
+def exportToPNMLStr(net:LabelledPetriNet) -> ET.Element: 
```
should be a function of the structure

adamburkegh replied:
nope

AdamBanham commented on pmkoalas/models/petrinet.py:
```diff
@@ -0,0 +1,341 @@
+    ET.indent( xml ) 
+    return ET.tostring(xml,encoding=ENCODING)
+
+def exportToPNML(net:LabelledPetriNet,fname:str): 
```
Should be a function of the structure.

adamburkegh replied:
nope

AdamBanham commented on pmkoalas/models/pnformatter.py:
```diff
@@ -0,0 +1,79 @@
+from pmkoalas.models.petrinet import PetriNetDOTFormatter
+from pmkoalas.models.dotutil import exportDOTToImage
+
+
```
So you have a separate file for half of the formatting code? Do you want to move all formatting code to this function or do you want to move this code to petrinet?

adamburkegh replied:
Definitely should be separate to petrinet

AdamBanham commented on tests/models/test_petrinet.py:
```diff
@@ -0,0 +1,77 @@
+  </net>
+</pnml>
+'''
+
```
This seems light for a major structure. Perhaps we should have a few more functional checks for this structure?

adamburkegh replied:
Requires changing the export implementation to be strictly deterministic by implementing sort comparators for places, transitions and arcs. Perhaps desirable, but I deliberately cut from the scope.

adamburkegh replied:
If you are talking about PN itself, the coverage provided by pnfrag is already substantial, as the comment notes.

adamburkegh commented:
> Hi Adam,
> 
> Could you provide some code snippets of how one might create a "Petri net"?
> 
> For example could you finish the following:
> 
> ```python
> from pmkoalas.dtlog import convert
> from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
> from pmkoalas.models.petrinet import LabelledPetriNet
> 
> # create a log
> log = convert( 
>     "a b e d z",
>     "a c e f z",
>     "a y e g z"
> )
> 
> # run alpha miner 
> miner = AlphaMinerInstance()
> places, transitions, flows = miner.mine_model(log)
> 
> # use given places, transitions and flows to create an 
> # instance of a petri net 
> net = LabelledPetriNet(places, transitions, flows)
> # display net in some potential human readable manner
> print(net)
> ```
> 
> However, running the above on the fork produces the following:
> 
> ```
> Traceback (most recent call last):
>   File "F:\Projects\Koalas\working\net_test.py", line 20, in <module>
>     print(net)
>   File "F:\Projects\Koalas\pmkoalas\models\petrinet.py", line 59, in __str__
>     return "LabelledPetriNet(" + self._label + ")\n" \
> TypeError: can only concatenate str (not "NoneType") to str
> ```

See `PetriNetFragmentTest.test_ptp_fragment()`

adamburkegh commented:
> Also I checked out the code, it seems mostly uncommented. We should at least strive to have the key entry points with a clear indication of how to use the function call.

Unit tests are better documentation than comments for new unstable APIs.

adamburkegh commented:
> Also I checked out the code, it seems mostly uncommented. We should at least strive to have the key entry points with a clear indication of how to use the function call.

Unit tests are better documentation than comments for new unstable APIs.

AdamBanham commented:
> > Also I checked out the code, it seems mostly uncommented. We should at least strive to have the key entry points with a clear indication of how to use the function call.
> 
> Unit tests are better documentation than comments for new unstable APIs.

Adam, the tests are not in the module on pypi. You need to add a decent comment on the structure and its initial call.

AdamBanham commented:
> > Hi Adam,
> > Could you provide some code snippets of how one might create a "Petri net"?
> > For example could you finish the following:
> > ```python
> > from pmkoalas.dtlog import convert
> > from pmkoalas.discovery.alpha_miner import AlphaMinerInstance
> > from pmkoalas.models.petrinet import LabelledPetriNet
> > 
> > # create a log
> > log = convert( 
> >     "a b e d z",
> >     "a c e f z",
> >     "a y e g z"
> > )
> > 
> > # run alpha miner 
> > miner = AlphaMinerInstance()
> > places, transitions, flows = miner.mine_model(log)
> > 
> > # use given places, transitions and flows to create an 
> > # instance of a petri net 
> > net = LabelledPetriNet(places, transitions, flows)
> > # display net in some potential human readable manner
> > print(net)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, running the above on the fork produces the following:
> > ```
> > Traceback (most recent call last):
> >   File "F:\Projects\Koalas\working\net_test.py", line 20, in <module>
> >     print(net)
> >   File "F:\Projects\Koalas\pmkoalas\models\petrinet.py", line 59, in __str__
> >     return "LabelledPetriNet(" + self._label + ")\n" \
> > TypeError: can only concatenate str (not "NoneType") to str
> > ```
> 
> See `PetriNetFragmentTest.test_ptp_fragment()`

This test only makes more questions than answers. However, it is clear that the structure is not a labelled Petri net at all. Should we enforce the elements of each set? Should we enforce that sets be passed? Nor does this address that the given default values are a valid configuration, as the error above notes.

adamburkegh commented:
> > > Also I checked out the code, it seems mostly uncommented. We should at least strive to have the key entry points with a clear indication of how to use the function call.
> > 
> > 
> > Unit tests are better documentation than comments for new unstable APIs.
> 
> Adam, the tests are not in the module on pypi. You need to add a decent comment on the structure and its initial call.

It's not necessary for a version 0.01 and is creating unnecessary maintenance overhead at a time when interfaces are very fluid. All that said, it's not a hard point of disagreement for me. Since you feel very strongly about it, I can add a few more comments.

AdamBanham commented on pmkoalas/models/pnfrag.py:
```diff
@@ -0,0 +1,344 @@
+    def readd_arc(self,fromNode,toNode):
+        self.net.add_arc_between(fromNode,toNode)
+
+    def nextId(self,newfloor=None):
+        if newfloor:
+            self.ctid = max(newfloor,ctid)
+        self.ctid += 1
+        return self.ctid
```
Is this not a bug? `ctid` in this context is undefined, it has not been set at all. I can see how `self.ctid` could be not but the non-self reference on 341 in the function call.

adamburkegh replied:
Looks like dead code, probably how the bug persisted. Have fixed in latest commit.

AdamBanham commented on pmkoalas/models/pnfrag.py:
```diff
@@ -0,0 +1,344 @@
+class ParseException(Exception):
+    pass
+
+def createNet(label, netText) -> LabelledPetriNet:
+    PetriNetFragmentParser().createNet(label,netText)
```
This is the entry point for the user right?

AdamBanham commented:
I found the entry point to petri net fragment somewhat difficult to follow. Before tackling the naming conventions (which I might leave alone for the pnfrag file), I wanted to see if you would allow an alternative function call to quickly consume fragments.

Based on the last commit, check out the following code:
```python
import pmkoalas.models.petrinet as pn 
import pmkoalas.models.pnfrag as pnfrag

# test exporting on simple net
net = pn.LabelledPetriNet(
          places=[
            pn.Place("p1",pid="f9da4b8b-2ebe-4384-a2f1-ae484cf898c8"),
            pn.Place("p2",pid="1da99779-a974-40b3-9573-bf34ab4c43cc"),
          ],
          transitions=[
            pn.Transition("t1",tid="24747a8f-0f98-4f2a-9694-1a6687c01334",
                       weight=1.0,silent=False),
          ],
          arcs=[
            pn.Arc(fromNode=pn.Place("p1",pid="f9da4b8b-2ebe-4384-a2f1-ae484cf898c8"),
                toNode=pn.Transition("t1",tid="24747a8f-0f98-4f2a-9694-1a6687c01334",
                                  weight=1.0,silent=False)
            ),
            pn.Arc(fromNode=pn.Transition("t1",tid="24747a8f-0f98-4f2a-9694-1a6687c01334",
                                    weight=1.0,silent=False),
                toNode=pn.Place("p1",pid="f9da4b8b-2ebe-4384-a2f1-ae484cf898c8")),
          ],
          name='Petri net'
        )
print(net)
pn.export_net_to_pnml(net, "./testing_import.pnml")
dotcode = pn.convert_net_to_dot(net)
with open("./test_dot.dot", "w") as f:
    f.write(dotcode)

# build a larger model
parser = pnfrag.PetriNetFragmentParser()
onet = parser.createNet("tester_net", "I -> [A] -> p1 -> [tau] -> F")
parser.addToNet(onet, "I -> [A] -> p1 -> [B] -> p2 -> [E] -> p3 -> [G] -> F")
parser.addToNet(onet, "I -> [A] -> p1 -> [C] -> p2 -> [E] -> p3 -> [H] -> F")
parser.addToNet(onet, "I -> [A] -> p1 -> [D] -> p2 -> [E] -> p3 -> [K] -> F")
onet = onet.create_net()
print(onet)
## alternative build pattern for fragments
net = pnfrag.parse_net_fragments("tester_net",
    "I -> [A] -> p1 -> [tau] -> F",
    "I -> [A] -> p1 -> [B] -> p2 -> [E] -> p3 -> [G] -> F",
    "I -> [A] -> p1 -> [C] -> p2 -> [E] -> p3 -> [H] -> F",
    "I -> [A] -> p1 -> [D] -> p2 -> [E] -> p3 -> [K] -> F"         
)
print(f"equality test : {onet==net=}")
pn.export_net_to_pnml(net, "./testing_import_2.pnml")
dotcode = pn.convert_net_to_dot(net)
with open("./test_dot_2.dot", "w") as f:
    f.write(dotcode)
```

adamburkegh commented:
> I found the entry point to petri net fragment somewhat difficult to follow. Before tackling the naming conventions (which I might leave alone for the pnfrag file), I wanted to see if you would allow an alternative function call to quickly consume fragments.
> 
> Based on the last commit, check out the following code:
> 


> onet = onet.create_net()
> net = pnfrag.parse_net_fragments("tester_net",
>     "I -> [A] -> p1 -> [tau] -> F",
>     "I -> [A] -> p1 -> [B] -> p2 -> [E] -> p3 -> [G] -> F",
>     "I -> [A] -> p1 -> [C] -> p2 -> [E] -> p3 -> [H] -> F",
>     "I -> [A] -> p1 -> [D] -> p2 -> [E] -> p3 -> [K] -> F"         
> )

I think convenience function entry points are a good idea. I added some tests and fixed the style convention to fit the rest of the module. I removed the repetition of the grammar in the doc comment, as it is a technical spec that needs to be maintained in a single place.

adamburkegh commented:
At what point do we merge out of this long-lived branch and do the rest of the tinkering in the main project?

AdamBanham commented:
> 



> At what point do we merge out of this long-lived branch and do the rest of the tinkering in the main project?

I think it is ready to merge, did you want to shift the version number and make a new dist?

adamburkegh commented:
I will commit the version change separately
